devito/types/dense.py

```diff
@@ -1454,3 +1453,160 @@ def parent(self):
         return self._parent
 
     _pickle_kwargs = Function._pickle_kwargs + ['parent']
+
+
+class TempFunction(DiscreteFunction):
+
+    """
+    Tensor symbol used to store an intermediate sub-expression extracted from
+    one or more symbolic equations.
+
+    Users should not instantiate this class directly. TempFunctions may be created
+    by Devito to store intermediate sub-expressions ("temporary values") when the
+    user supplies the `cire-ftemps` option to an Operator.
+
+    Unlike other DiscreteFunction types, TempFunctions do not carry data directly.
+    However, they can generate Functions to override the TempFunction at Operator
+    application time (see the Examples section below).
+
+    TempFunctions are useful if the user wants to retain control over the allocation
+    and deletion of temporary storage (by default, instead, Devito uses Arrays, which
+    are allocated and deallocated upon entering and exiting C-land, respectively).
+
+    Examples
+    --------
+    The `make` method makes the TempFunction create a new Function. For more info,
+    refer to TempFunction.make.__doc__.
+
+      .. code-block:: python
+
+        op = Operator(...)
+        cfuncs = [i for i in op.input if i.is_TempFunction]
+        kwargs = {i.name: i.make(grid.shape) for i in cfuncs}
+        op.apply(..., **kwargs)
+    """
+
+    is_TempFunction = True
+
+    def __init_finalize__(self, *args, **kwargs):
+        super().__init_finalize__(*args, **kwargs)
+
+        self._pointer_dim = kwargs.get('pointer_dim')
+
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
+        pointer_dim = kwargs.get('pointer_dim')
+        dimensions = as_tuple(kwargs['dimensions'])
+        if pointer_dim not in dimensions:
```

Comment:
`as_tuple(set(...))`

Reply:
?

above is tuplyfied already?

Reply:
I mean instead of if can just do `dimensions = tuple(set(dimensions, pointer_dim))`

Reply:
I cannot use set or I'll lose the ordering